### ISSUE TITLE
No heatmap shown with `seaborn` and `%matplotlib widget`

### ISSUE THREAD

**Issue by dominiquesydow** [matplotlib/ipympl#391]
## Describe the issue

Thank you very much for providing this tool!

I have an apparently `seaborn`-related problem with showing widgets; I am posting this here first. Please let me know if you think this problem is rather on `seaborn`'s end and I'll repost it there.

---

The problem: `seaborn` figures are not shown in JupyterLab with `%matplotlib widget` (they do show up with `%matplotlib inline`):

```
%matplotlib widget
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

df = pd.DataFrame([[10, 20, 30, 40], [50, 30, 8, 15],
                   [25, 14, 41, 8], [7, 14, 21, 28]])
sns.heatmap(df, cmap='RdYlGn', linewidths=0.30, annot=True)
```
![image](https://user-images.githubusercontent.com/7207093/141105912-278b891b-14fd-447a-8205-87e9851c0f64.png)

```
%matplotlib inline
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

df = pd.DataFrame([[10, 20, 30, 40], [50, 30, 8, 15],
                   [25, 14, 41, 8], [7, 14, 21, 28]])
sns.heatmap(df, cmap='RdYlGn', linewidths=0.30, annot=True)
```
![image](https://user-images.githubusercontent.com/7207093/141107251-32aaa188-d6e9-4505-b8bb-0e8f539e4294.png)

---

No problem with `matplotlib`:

```
%matplotlib widget
import matplotlib.pyplot as plt

plt.plot([[1, 2], [0, 0]])
```
![image](https://user-images.githubusercontent.com/7207093/141106050-b2e8d6ba-0c58-4ff8-9d33-890909d06e1d.png)

## Versions

```
ipympl version: 0.8.2
Selected Jupyter core packages...
IPython          : 7.29.0
ipykernel        : 6.4.2
ipywidgets       : 7.6.5
jupyter_client   : 7.0.6
jupyter_core     : 4.9.1
jupyter_server   : 1.11.2
jupyterlab       : 3.2.2
nbclient         : 0.5.5
nbconvert        : 6.2.0
nbformat         : 5.1.3
notebook         : 6.4.5
qtconsole        : not installed
traitlets        : 5.1.1
Known nbextensions:
  config dir: /home/dominique/.jupyter/nbconfig
    notebook section
      nbextensions_configurator/config_menu/main  enabled 
      - Validating: problems found:
        - require?  X nbextensions_configurator/config_menu/main
      contrib_nbextensions_help_item/main  enabled 
      - Validating: OK
      spellchecker/main  enabled 
      - Validating: OK
      gist_it/main  enabled 
      - Validating: OK
    tree section
      nbextensions_configurator/tree_tab/main  enabled 
      - Validating: problems found:
        - require?  X nbextensions_configurator/tree_tab/main
  config dir: /home/dominique/.local/miniconda/envs/test/etc/jupyter/nbconfig
    notebook section
      jupyter-matplotlib/extension  enabled 
      - Validating: OK
      jupyter-js-widgets/extension  enabled 
      - Validating: OK
JupyterLab v3.2.2
/home/dominique/.local/miniconda/envs/test/share/jupyter/labextensions
        jupyter-matplotlib v0.10.2 enabled OK
        @jupyter-widgets/jupyterlab-manager v3.0.1 enabled OK (python, jupyterlab_widgets)

```

Installed from conda-forge
```
mamba create -n test matplotlib pandas seaborn jupyterlab ipympl -y -c conda-forge
```

More versions:
```
# Name                    Version                   Build  Channel
matplotlib                3.4.3            py39hf3d152e_1    conda-forge
matplotlib-base           3.4.3            py39h2fa2bec_1    conda-forge
matplotlib-inline         0.1.3              pyhd8ed1ab_0    conda-forge
pandas                    1.3.4            py39hde0f152_1    conda-forge
seaborn                   0.11.2               hd8ed1ab_0    conda-forge
seaborn-base              0.11.2             pyhd8ed1ab_0    conda-forge
```

Do you have an idea why this is happening?

**Comment by ianhi** [comment]
This worked in ipympl 0.8.0 but not in `0.8.2` so this may be something to do with the new display logic. Although interestingly just calling `plt.gca` (which is what seaborn does) seems to work.

**Comment by ianhi** [comment]
@mwaskom do you have any insight? it seems like you just call `plt.gca()` but I suppose something that's happening in `plotter.plot` is making things go poorly.

**Comment by mwaskom** [comment]
I think I'd want to know a few things

- Is this a "seaborn" problem or a "heatmap" problem?
- If the latter, does a reasonably similar plot made only in matplotlib (i.e., pcolormesh with some texts on it) work?

**Comment by dominiquesydow** [comment]
Thanks a lot for getting back to me so quickly!
Let me try to apply your suggestions.

---
> it seems like you just call `plt.gca()` but I suppose something that's happening in `plotter.plot` is making things go poorly.

Do you mean like this?
![image](https://user-images.githubusercontent.com/7207093/141267961-44eedac0-f3de-453f-bf42-d5d103083620.png)

---
> * If the latter, does a reasonably similar plot made only in matplotlib (i.e., pcolormesh with some texts on it) work?

Yes, it seems that works.
![image](https://user-images.githubusercontent.com/7207093/141267492-d9f59ecf-da75-4951-bfa9-897e28dc000e.png)

---
> * If you can make a similar matplotlib plot with no issues, can you get the heatmap to work if you boil it down to the simplest version of what seaborn can draw? (i.e. no annotations, no colorbar, and disable the automatic selection of tick label density)
![image](https://user-images.githubusercontent.com/7207093/141268072-33c5e72f-8b73-4510-9715-3058652b3d0b.png)

The same behaviour with
```python
sns.heatmap(df, xticklabels=False, yticklabels=False)
```
and
```python
sns.heatmap(df, xticklabels=True, yticklabels=True)
```
assuming this is what was meant with
> disable the automatic selection of tick label density

**Comment by mwaskom** [comment]
Yes that's right about the tick labels. But you could also try isolating the colorbar: it's on by default in the seaborn plot, so you could try turning it off there (`cbar=False` and turning it on for the matplotlib plot (`plt.colorbar`).

But I think that doesn't answer the other question which is ... can you use _other_ seaborn functions, or is this an issue specific to `heatmap`?

**Comment by ianhi** [comment]
A crucial thing in these tests is probably to always start with `plt.close('all')` so that the behavior of `plt.gca` is consistent

**Comment by ianhi** [comment]
Also @dominiquesydow can you please post the code snippets as text so that they are copy-pasteable - thanks!

**Comment by dominiquesydow** [comment]
Hi @ianhi,

> Also @dominiquesydow can you please post the code snippets as text so that they are copy-pasteable - thanks!

I have updated my comment with copy-pasteable code.

Also, I have tried to investigate your initial question (my apologies for not getting to at first)

> * Is this a "seaborn" problem or a "`seaborn.heatmap`" problem?

I set up [this notebook](https://gist.github.com/dominiquesydow/b706b076fbf0495300342d7f431f493a); below a summary: 

- Other `seaborn` plots seem fine to me.
- I tried two versions that show the same behaviour
  - (1) Define figure and axes (`fig, ax = plt.subplots(nrows=1, ncols=1)`) and pass axis to `seaborn` functions (`ax` parameter)
  - (2) Close all axes and then get current axis (`plt.close('all')` followed by `ax = plt.gca()`)
- Interesting observation: Maybe the `seaborn` heatmap tries to draw to the axis _before_ the current axis (see the end of the posted notebook)
  - Procedure: 1. I draw e.g. a scatter plot, 2. I draw heatmap _without_ clearing axes
  - Result: The heatmap is drawn on top of the previous plot.
- Note: I see the same behaviour with Jupyter Lab and Jupyter Notebook
- Note: I do not see any plots when running all cells at once; I have to run cell-by-cell

**Comment by mwaskom** [comment]
That behavior all looks correct to me. If you don't provide seaborn with an `Axes` target, it plots on the "current" Axes (creating it if necessary). The Jupyter inline calls `plt.close` behind the scenes after it executes every cell, but `ipyml` doesn't (so e.g. you can create a plot in one cell then update it by doing someone in another one).

**Comment by 01baftb** [comment]
I am having similar issue https://github.com/matplotlib/ipympl/issues/60#issuecomment-973722481

**Comment by dominiquesydow** [comment]
Hi @mwaskom,

Thanks for getting back to me. 
I am showing a minimal example here to show what is still confusing me.

```python
# First cell
%matplotlib widget
tips = sns.load_dataset("tips")

fig, ax = plt.subplots(nrows=1, ncols=1)
sns.scatterplot(data=tips, x="total_bill", y="tip", hue="time", ax=ax)
```
This works nicely, I get the plot:
![image](https://user-images.githubusercontent.com/7207093/142579930-856d548e-56b4-441a-8faf-a5f0ae77cf9d.png)

Then, I call `plt.close` as you suggested:

```python
# Second cell
plt.close()
```

Then, I try to plot `sns.heatmap` and get an empty plot:

```python
# Third cell
import numpy as np

uniform_data = np.random.rand(10, 12)

fig, ax = plt.subplots(nrows=1, ncols=1)
sns.heatmap(uniform_data, ax=ax)
```

![image](https://user-images.githubusercontent.com/7207093/142580213-0743dfd7-2b2c-4fc0-869d-9393c1ffca28.png)

Do you have an idea on how to make the last plot appear? 

> If you don't provide seaborn with an Axes target, it plots on the "current" Axes (creating it if necessary). The Jupyter inline backend calls plt.close behind the scenes after it executes every cell, but ipyml doesn't (so e.g. you can create a plot in one cell then update it by doing something in another one).

With `ax=ax`, I am providing `sns.heatmap` with an Axes target (at least I think I am); I was expecting `ipympl` to use that Axes as well.

**Comment by mwaskom** [comment]
I'm confused about how what you're describing there is not represented in the notebook you shared? You seem to be plotting `heatmap` just fine there?

**Comment by dominiquesydow** [comment]
My apologies, apparently uploading the notebook to GH makes the heatmap appear... running the same notebook locally shows the behaviour discussed [in this comment](https://github.com/matplotlib/ipympl/issues/391#issuecomment-973816424).

For the time being, I will pin `ipympl` to 0.8.0; here the `seaborn` heatmaps still appear as expected.

**Comment by mwaskom** [comment]
Oh that's interesting! I'd say this sounds like an issue on the ipympl side of things. I wish I could offer a hypothesis based on the fact that it seems to happen with heatmaps, but nothing comes immediately to mind beyond what I suggested above.

**Comment by dominiquesydow** [comment]
Thanks for taking the time to discuss this issue with me. 

Although I cannot offer a solution with the latest `ipympl` version, downgrading to 0.8.0 helped; I am therefore closing this issue. Please re-open if more discussion is needed for a solution for the latest version.

**Comment by ianhi** [comment]
> Please re-open if more discussion is needed for a solution for the latest version.

Yes this is definitely a bug than needs fixing - thanks for finding it!
